api/server.py

```diff
@@ -180,35 +180,35 @@ def expression():
         response = database_lib.expression(body).to_list()
         return jsonify(response)
 
-@app.route('/api/content.json', methods=['POST', 'GET'])
-def content():
-    if request.method == 'GET':
-        with open('contents.json') as json_file:
-            response = json.load(json_file)
-        return jsonify(response)
-    elif request.method == 'POST':
-        if not is_whitelisted():
-            return redirect(url_for("login"))
-        else:
-            body = request.json
-            with open('contents.json', 'w') as json_file:
-                json.dump(body, json_file, indent=2)
-            return jsonify({"success": True})
-
-@app.route('/api/content.staged.json', methods=['POST', 'GET'])
-def content_staged():
-    if request.method == 'GET':
-        with open('contents.staged.json') as json_file:
-            response = json.load(json_file)
-        return jsonify(response)
-    elif request.method == 'POST':
-        if not is_whitelisted():
-            return jsonify({"success": False})
-        else:
-            body = request.json
-            with open('contents.staged.json', 'w') as json_file:
-                json.dump(body, json_file, indent=2)
-            return jsonify({"success": True})
+content_files = [
+    'content.json',
+    'content.staged.json',
+]
+
+import tempfile
+import shutil
+
+def add_content_route(content_file):
+    endpoint = '/api/' + content_file
+    @app.route(endpoint, methods=['POST', 'GET'], endpoint=endpoint)
+    def content():
+        if request.method == 'GET':
+            with open(content_file) as json_file:
+                response = json.load(json_file)
+            return jsonify(response)
+        elif request.method == 'POST':
+            if not is_whitelisted():
+                return redirect(url_for("login"))
+            else:
+                body = request.json
+                with tempfile.NamedTemporaryFile(mode='w') as fp:
+                    json.dump(body, fp, indent=2)
+                    fp.flush()
+                    shutil.copy2(fp.name, content_file)
```

Comment:
This was more annoying to do than expected, but is this enough @pontus ?

Reply:
I'd normally be more comfortable with `delete=False` (and `dir` pointing to the correct directory so the tempfile is created on the same file system) to `NamedTemporaryFile` and `fp.close()` followed by `os.rename`/`shutil.move` instead.

I do not see any promises in the documentation for `shutil` that things are visible as atomic, which is the technical complain on this solution.

Reply:
Alright. Is `os.rename` what you recommend then? Do you know if `os.rename` will be atomic on linux?

Reply:
(Using `delete=False`)

Reply:
I'm more used to using the functions from the `os` package, but either would be fine (if you are more comfortable with `shutil`.

We had a discussion some time ago, but in short, typically on unixes, `os.rename` within the same filesystem (e.g. in the same directory) will be atomic to outside observators.